File: skills/investment-analyzer/scripts/valuation_engine.py
#!/usr/bin/env python3
"""
Valuation Engine - 估值模型计算模块

估值原则：
1. 多模型交叉验证
2. 敏感性分析
3. 明确假设条件
4. 估值区间表达
"""

import json
import sys
from typing import Dict, List, Optional, Tuple, Any
from dataclasses import dataclass
from datetime import datetime
import math


@dataclass
class ValuationResult:
    """估值结果数据结构"""
    method: str
    value: float
    assumptions: Dict
    sensitivity: Dict
    confidence: str  # high, medium, low


class ValuationEngine:
    """估值引擎"""
    
    def __init__(self):
        self.warnings = []
        self.risk_free_rate = 0.04  # 无风险利率假设
        self.market_risk_premium = 0.06  # 市场风险溢价假设
    
    def _validate_input(self, value: float, name: str, 
                       min_val: Optional[float] = None,
                       max_val: Optional[float] = None) -> float:
        """验证输入数据"""
        if value is None or math.isnan(value):
            raise ValueError(f"{name}不能为空或NaN")
        
        if min_val is not None and value < min_val:
            self.warnings.append(f"{name}({value})低于最小值{min_val}")
        
        if max_val is not None and value > max_val:
            self.warnings.append(f"{name}({value})超过最大值{max_val}")
        
        return float(value)
    
    def calculate_pe_valuation(self, data: Dict) -> Dict:
        """
        市盈率估值法
        
        Args:
            data: {
                "eps": 每股收益,
                "pe_ratio": 当前PE,
                "target_pe": 目标PE,
                "growth_rate": 预期增长率(%)
            }
            
        Returns:
            PE估值结果
        """
        eps = self._validate_input(data.get('eps', 0), "EPS", min_val=0.01)
        current_pe = self._validate_input(data.get('pe_ratio', 0), "当前PE", min_val=0)
        target_pe = data.get('target_pe', current_pe)
        growth_rate = data.get('growth_rate', 0)
        
        # 基于目标PE计算目标价
        target_price = eps * target_pe
        
        # PEG分析
        peg = current_pe / growth_rate if growth_rate > 0 else float('inf')
        
        # 敏感性分析
        sensitivity = {}
        for pe in [target_pe * 0.8, target_pe, target_pe * 1.2]:
            sensitivity[f"PE_{pe:.1f}"] = round(eps * pe, 2)
        
        return {
            "method": "PE估值法",
            "target_price": round(target_price, 2),
            "current_pe": round(current_pe, 2),
            "target_pe": round(target_pe, 2),
            "peg": round(peg, 2) if peg != float('inf') else None,
            "peg_assessment": "低估" if peg < 1 else "合理" if peg <= 2 else "高估",
            "sensitivity": sensitivity,
            "assumptions": {
                "eps": eps,
                "target_pe": target_pe,
                "growth_rate": f"{growth_rate}%"
            },
            "confidence": "medium"
        }
    
    def calculate_pb_valuation(self, data: Dict) -> Dict:
        """
        市净率估值法
        
        Args:
            data: {
                "book_value_per_share": 每股净资产,
                "pb_ratio": 当前PB,
                "target_pb": 目标PB,
                "roe": ROE(%)
            }
        """
        bvps = self._validate_input(data.get('book_value_per_share', 0), "每股净资产", min_val=0.01)
        current_pb = self._validate_input(data.get('pb_ratio', 0), "当前PB", min_val=0)
        target_pb = data.get('target_pb', current_pb)
        roe = data.get('roe', 0)
        
        # 基于目标PB计算目标价
        target_price = bvps * target_pb
        
        # 合理PB估算 (基于ROE)
        # 合理PB ≈ ROE / 要求回报率
        required_return = self.risk_free_rate + self.market_risk_premium
        fair_pb = (roe / 100) / required_return if roe > 0 else 0
        
        # 敏感性分析
        sensitivity = {}
        for pb in [target_pb * 0.8, target_pb, target_pb * 1.2]:
            sensitivity[f"PB_{pb:.1f}"] = round(bvps * pb, 2)
        
        return {
            "method": "PB估值法",
            "target_price": round(target_price, 2),
            "current_pb": round(current_pb, 2),
            "target_pb": round(target_pb, 2),
            "fair_pb": round(fair_pb, 2),
            "pb_premium": round((current_pb / fair_pb - 1) * 100, 2) if fair_pb > 0 else None,
            "sensitivity": sensitivity,
            "assumptions": {
                "book_value_per_share": bvps,
                "target_pb": target_pb,
                "roe": f"{roe}%"
            },
            "confidence": "medium"
        }
    
    def calculate_dcf_valuation(self, data: Dict) -> Dict:
        """
        自由现金流折现模型
        
        Args:
            data: {
                "fcf": 当前自由现金流,
                "growth_rate_5y": 前5年增长率(%),
                "growth_rate_terminal": 永续增长率(%),
                "wacc": 加权平均资本成本(%),
                "shares_outstanding": 总股本,
                "net_debt": 净负债
            }
        """
        fcf = self._validate_input(data.get('fcf', 0), "FCF")
        growth_5y = self._validate_input(data.get('growth_rate_5y', 0), "5年增长", min_val=-50, max_val=100)
        growth_terminal = self._validate_input(data.get('growth_rate_terminal', 2), "永续增长", min_val=0, max_val=5)
        wacc = self._validate_input(data.get('wacc', 10), "WACC", min_val=1, max_val=30)
        shares = self._validate_input(data.get('shares_outstanding', 1), "总股本", min_val=1)
        net_debt = data.get('net_debt', 0)
        
        # 转换为小数
        g5 = growth_5y / 100
        gt = growth_terminal / 100
        r = wacc / 100
        
        # 计算预测期现金流现值
        pv_fcf = 0
        fcf_list = []
        for year in range(1, 6):
            fcf_year = fcf * ((1 + g5) ** year)
            pv = fcf_year / ((1 + r) ** year)
            pv_fcf += pv
            fcf_list.append({
                "year": year,
                "fcf": round(fcf_year, 2),
                "pv": round(pv, 2)
            })
        
        # 计算终值
        fcf_year5 = fcf * ((1 + g5) ** 5)
        terminal_value = fcf_year5 * (1 + gt) / (r - gt)
        pv_terminal = terminal_value / ((1 + r) ** 5)
        
        # 企业价值
        enterprise_value = pv_fcf + pv_terminal
        
        # 股权价值
        equity_value = enterprise_value - net_debt
        
        # 每股价值
        value_per_share = equity_value / shares
        
        # 敏感性分析 (WACC vs 增长率)
        sensitivity = {}
        for w in [r - 0.02, r, r + 0.02]:
            for g in [g5 - 0.05, g5, g5 + 0.05]:
                if w > gt:
                    pv_fcf_sens = sum(fcf * ((1 + g) ** y) / ((1 + w) ** y) for y in range(1, 6))
                    fcf5 = fcf * ((1 + g) ** 5)
                    tv = fcf5 * (1 + gt) / (w - gt)
                    pv_tv = tv / ((1 + w) ** 5)
                    ev = pv_fcf_sens + pv_tv
                    eq = ev - net_debt
                    price = eq / shares
                    sensitivity[f"WACC{w*100:.0f}%_G{g*100:.0f}%"] = round(price, 2)
        
        return {
            "method": "DCF估值法",
            "value_per_share": round(value_per_share, 2),
            "enterprise_value": round(enterprise_value, 2),
            "equity_value": round(equity_value, 2),
            "pv_fcf_5y": round(pv_fcf, 2),
            "pv_terminal": round(pv_terminal, 2),
            "terminal_value": round(terminal_value, 2),
            "fcf_projection": fcf_list,
            "sensitivity": sensitivity,
            "assumptions": {
                "current_fcf": fcf,
                "growth_rate_5y": f"{growth_5y}%",
                "growth_rate_terminal": f"{growth_terminal}%",
                "wacc": f"{wacc}%",
                "shares_outstanding": shares,
                "net_debt": net_debt
            },
            "confidence": "medium"
        }
    
    def calculate_evebitda_valuation(self, data: Dict) -> Dict:
        """
        EV/EBITDA估值法
        
        Args:
            data: {
                "ebitda": EBITDA,
                "net_debt": 净负债,
                "shares_outstanding": 总股本,
                "ev_ebitda_current": 当前EV/EBITDA,
                "ev_ebitda_target": 目标EV/EBITDA
            }
        """
        ebitda = self._validate_input(data.get('ebitda', 0), "EBITDA", min_val=0.01)
        net_debt = data.get('net_debt', 0)
        shares = self._validate_input(data.get('shares_outstanding', 1), "总股本", min_val=1)
        current_multiple = self._validate_input(data.get('ev_ebitda_current', 0), "当前EV/EBITDA", min_val=0)
        target_multiple = data.get('ev_ebitda_target', current_multiple)
        
        # 计算目标企业价值
        target_ev = ebitda * target_multiple
        
        # 股权价值
        equity_value = target_ev - net_debt
        
        # 每股价值
        value_per_share = equity_value / shares
        
        # 敏感性分析
        sensitivity = {}
        for mult in [target_multiple * 0.8, target_multiple, target_multiple * 1.2]:
            ev = ebitda * mult
            eq = ev - net_debt
            price = eq / shares
            sensitivity[f"EV/EBITDA_{mult:.1f}x"] = round(price, 2)
        
        return {
            "method": "EV/EBITDA估值法",
            "value_per_share": round(value_per_share, 2),
            "enterprise_value": round(target_ev, 2),
            "equity_value": round(equity_value, 2),
            "current_multiple": round(current_multiple, 2),
            "target_multiple": round(target_multiple, 2),
            "sensitivity": sensitivity,
            "assumptions": {
                "ebitda": ebitda,
                "net_debt": net_debt,
                "target_multiple": target_multiple
            },
            "confidence": "medium"
        }
    
    def calculate_ddm_valuation(self, data: Dict) -> Dict:
        """
        股利贴现模型
        
        Args:
            data: {
                "dividend_per_share": 每股股利,
                "growth_rate": 股利增长率(%),
                "required_return": 要求回报率(%)
            }
        """
        dps = self._validate_input(data.get('dividend_per_share', 0), "每股股利", min_val=0)
        growth = self._validate_input(data.get('growth_rate', 2), "增长率", min_val=0, max_val=10)
        required = self._validate_input(data.get('required_return', 10), "要求回报率", min_val=growth + 1, max_val=30)
        
        g = growth / 100
        r = required / 100
        
        # Gordon增长模型
        if r > g:
            value = dps * (1 + g) / (r - g)
        else:
            value = 0
            self.warnings.append("要求回报率必须大于增长率")
        
        # 敏感性分析
        sensitivity = {}
        for req in [required - 2, required, required + 2]:
            for gr in [max(0, growth - 1), growth, growth + 1]:
                if req > gr:
                    val = dps * (1 + gr/100) / (req/100 - gr/100)
                    sensitivity[f"R{req:.0f}%_G{gr:.0f}%"] = round(val, 2)
        
        return {
            "method": "DDM股利贴现模型",
            "value_per_share": round(value, 2),
            "dividend_yield": round(dps / value * 100, 2) if value > 0 else 0,
            "sensitivity": sensitivity,
            "assumptions": {
                "dividend_per_share": dps,
                "growth_rate": f"{growth}%",
                "required_return": f"{required}%"
            },
            "confidence": "high" if dps > 0 else "low"
        }
    
    def comprehensive_valuation(self, data: Dict) -> Dict:
        """
        综合估值分析
        
        Args:
            data: 包含各估值方法所需参数的字典
            
        Returns:
            综合估值结果
        """
        results = []
        
        # PE估值
        if 'eps' in data and data['eps'] > 0:
            try:
                pe_result = self.calculate_pe_valuation(data)
                results.append(pe_result)
            except Exception as e:
                self.warnings.append(f"PE估值失败: {str(e)}")
        
        # PB估值
        if 'book_value_per_share' in data and data['book_value_per_share'] > 0:
            try:
                pb_result = self.calculate_pb_valuation(data)
                results.append(pb_result)
            except Exception as e:
                self.warnings.append(f"PB估值失败: {str(e)}")
        
        # DCF估值
        if 'fcf' in data and data['fcf'] != 0:
            try:
                dcf_result = self.calculate_dcf_valuation(data)
                results.append(dcf_result)
            except Exception as e:
                self.warnings.append(f"DCF估值失败: {str(e)}")
        
        # EV/EBITDA估值
        if 'ebitda' in data and data['ebitda'] > 0:
            try:
                ev_result = self.calculate_evebitda_valuation(data)
                results.append(ev_result)
            except Exception as e:
                self.warnings.append(f"EV/EBITDA估值失败: {str(e)}")
        
        # DDM估值
        if 'dividend_per_share' in data and data['dividend_per_share'] > 0:
            try:
                ddm_result = self.calculate_ddm_valuation(data)
                results.append(ddm_result)
            except Exception as e:
                self.warnings.append(f"DDM估值失败: {str(e)}")
        
        # 计算估值区间
        if results:
            values = [r['value_per_share'] for r in results if 'value_per_share' in r]
            if values:
                valuation_range = {
                    "low": round(min(values), 2),
                    "high": round(max(values), 2),
                    "median": round(sorted(values)[len(values)//2], 2),
                    "average": round(sum(values)/len(values), 2)
                }
            else:
                valuation_range = None
        else:
            valuation_range = None
        
        return {
            "individual_results": results,
            "valuation_range": valuation_range,
            "warnings": self.warnings,
            "timestamp": datetime.now().isoformat()
        }


def main():
    """命令行入口"""
    if len(sys.argv) < 2:
        print("Usage: python valuation_engine.py <command> [args...]")
        print("Commands:")
        print("  pe <json_file>        - PE估值")
        print("  pb <json_file>        - PB估值")
        print("  dcf <json_file>       - DCF估值")
        print("  ev_ebitda <json_file> - EV/EBITDA估值")
        print("  ddm <json_file>       - DDM估值")
        print("  comprehensive <json_file> - 综合估值")
        sys.exit(1)
    
    command = sys.argv[1]
    
    try:
        if len(sys.argv) < 3:
            print("Error: 需要JSON文件路径")
            sys.exit(1)
        
        with open(sys.argv[2], 'r') as f:
            data = json.load(f)
        
        engine = ValuationEngine()
        
        if command == "pe":
            result = engine.calculate_pe_valuation(data)
        elif command == "pb":
            result = engine.calculate_pb_valuation(data)
        elif command == "dcf":
            result = engine.calculate_dcf_valuation(data)
        elif command == "ev_ebitda":
            result = engine.calculate_evebitda_valuation(data)
        elif command == "ddm":
            result = engine.calculate_ddm_valuation(data)
        elif command == "comprehensive":
            result = engine.comprehensive_valuation(data)
        else:
            print(f"Unknown command: {command}")
            sys.exit(1)
        
        print(json.dumps(result, indent=2, ensure_ascii=False))
        
    except Exception as e:
        print(json.dumps({"error": str(e)}, indent=2))
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/utils/valuationEngine.ts
export interface ValuationInput {
  eps?: number;
  pe_ratio?: number;
  target_pe?: number;
  growth_rate?: number;
  book_value_per_share?: number;
  pb_ratio?: number;
  target_pb?: number;
  roe?: number;
  fcf?: number;
  growth_rate_5y?: number;
  growth_rate_terminal?: number;
  wacc?: number;
  shares_outstanding?: number;
  net_debt?: number;
  ebitda?: number;
  ev_ebitda_current?: number;
  ev_ebitda_target?: number;
  dividend_per_share?: number;
  required_return?: number;
}

export interface ValuationResult {
  method: string;
  target_price?: number;
  value_per_share?: number;
  assumptions: Record<string, any>;
  sensitivity: Record<string, number>;
  confidence: 'high' | 'medium' | 'low';
}

export class ValuationEngine {
  private warnings: string[] = [];
  private riskFreeRate = 0.04;
  private marketRiskPremium = 0.06;

  calculatePEValuation(data: Partial<ValuationInput>): ValuationResult {
    const eps = data.eps || 0;
    const currentPE = data.pe_ratio || 0;
    const targetPE = data.target_pe || currentPE;
    const growthRate = data.growth_rate || 0;

    const targetPrice = eps * targetPE;
    const peg = growthRate > 0 ? currentPE / growthRate : Infinity;

    const sensitivity: Record<string, number> = {};
    [targetPE * 0.8, targetPE, targetPE * 1.2].forEach(pe => {
      sensitivity[`PE_${pe.toFixed(1)}`] = Math.round(eps * pe * 100) / 100;
    });

    return {
      method: 'PE估值法',
      target_price: Math.round(targetPrice * 100) / 100,
      assumptions: {
        eps,
        target_pe: targetPE,
        growth_rate: `${growthRate}%`,
        peg: peg === Infinity ? null : Math.round(peg * 100) / 100,
        peg_assessment: peg < 1 ? '低估' : peg <= 2 ? '合理' : '高估'
      },
      sensitivity,
      confidence: 'medium'
    };
  }

  calculatePBValuation(data: Partial<ValuationInput>): ValuationResult {
    const bvps = data.book_value_per_share || 0;
    const currentPB = data.pb_ratio || 0;
    const targetPB = data.target_pb || currentPB;
    const roe = data.roe || 0;

    const targetPrice = bvps * targetPB;
    const requiredReturn = this.riskFreeRate + this.marketRiskPremium;
    const fairPB = roe > 0 ? (roe / 100) / requiredReturn : 0;

    const sensitivity: Record<string, number> = {};
    [targetPB * 0.8, targetPB, targetPB * 1.2].forEach(pb => {
      sensitivity[`PB_${pb.toFixed(1)}`] = Math.round(bvps * pb * 100) / 100;
    });

    return {
      method: 'PB估值法',
      target_price: Math.round(targetPrice * 100) / 100,
      assumptions: {
        book_value_per_share: bvps,
        target_pb: targetPB,
        roe: `${roe}%`,
        fair_pb: Math.round(fairPB * 100) / 100,
        pb_premium: fairPB > 0 ? Math.round((currentPB / fairPB - 1) * 10000) / 100 : null
      },
      sensitivity,
      confidence: 'medium'
    };
  }

  calculateDCFValuation(data: Partial<ValuationInput>): ValuationResult {
    const fcf = data.fcf || 0;
    const growth5Y = (data.growth_rate_5y || 0) / 100;
    const growthTerminal = (data.growth_rate_terminal || 0.02) / 100;
    const wacc = (data.wacc || 10) / 100;
    const shares = data.shares_outstanding || 1;
    const netDebt = data.net_debt || 0;

    let pvFCF = 0;
    const fcfList: { year: number; fcf: number; pv: number }[] = [];

    for (let year = 1; year <= 5; year++) {
      const fcfYear = fcf * Math.pow(1 + growth5Y, year);
      const pv = fcfYear / Math.pow(1 + wacc, year);
      pvFCF += pv;
      fcfList.push({ year, fcf: Math.round(fcfYear * 100) / 100, pv: Math.round(pv * 100) / 100 });
    }

    const fcfYear5 = fcf * Math.pow(1 + growth5Y, 5);
    const terminalValue = fcfYear5 * (1 + growthTerminal) / (wacc - growthTerminal);
    const pvTerminal = terminalValue / Math.pow(1 + wacc, 5);
    const enterpriseValue = pvFCF + pvTerminal;
    const equityValue = enterpriseValue - netDebt;
    const valuePerShare = equityValue / shares;

    const sensitivity: Record<string, number> = {};
    [wacc - 0.02, wacc, wacc + 0.02].forEach(w => {
      [growth5Y - 0.05, growth5Y, growth5Y + 0.05].forEach(g => {
        if (w > growthTerminal) {
          let pv = 0;
          for (let y = 1; y <= 5; y++) {
            pv += fcf * Math.pow(1 + g, y) / Math.pow(1 + w, y);
          }
          const fcf5 = fcf * Math.pow(1 + g, 5);
          const tv = fcf5 * (1 + growthTerminal) / (w - growthTerminal);
          const pvtv = tv / Math.pow(1 + w, 5);
          const ev = pv + pvtv;
          const eq = ev - netDebt;
          const price = eq / shares;
          sensitivity[`WACC${(w * 100).toFixed(0)}%_G${(g * 100).toFixed(0)}%`] = Math.round(price * 100) / 100;
        }
      });
    });

    return {
      method: 'DCF估值法',
      value_per_share: Math.round(valuePerShare * 100) / 100,
      assumptions: {
        current_fcf: fcf,
        growth_rate_5y: `${(growth5Y * 100).toFixed(0)}%`,
        growth_rate_terminal: `${(growthTerminal * 100).toFixed(0)}%`,
        wacc: `${(wacc * 100).toFixed(0)}%`,
        shares_outstanding: shares,
        net_debt: netDebt,
        pv_fcf_5y: Math.round(pvFCF * 100) / 100,
        pv_terminal: Math.round(pvTerminal * 100) / 100
      },
      sensitivity,
      confidence: 'medium'
    };
  }

  calculateEVEBITDAValuation(data: Partial<ValuationInput>): ValuationResult {
    const ebitda = data.ebitda || 0;
    const netDebt = data.net_debt || 0;
    const shares = data.shares_outstanding || 1;
    const currentMultiple = data.ev_ebitda_current || 0;
    const targetMultiple = data.ev_ebitda_target || currentMultiple;

    const targetEV = ebitda * targetMultiple;
    const equityValue = targetEV - netDebt;
    const valuePerShare = equityValue / shares;

    const sensitivity: Record<string, number> = {};
    [targetMultiple * 0.8, targetMultiple, targetMultiple * 1.2].forEach(mult => {
      const ev = ebitda * mult;
      const eq = ev - netDebt;
      const price = eq / shares;
      sensitivity[`EV/EBITDA_${mult.toFixed(1)}x`] = Math.round(price * 100) / 100;
    });

    return {
      method: 'EV/EBITDA估值法',
      value_per_share: Math.round(valuePerShare * 100) / 100,
      assumptions: {
        ebitda,
        net_debt: netDebt,
        target_multiple: targetMultiple,
        enterprise_value: Math.round(targetEV * 100) / 100,
        equity_value: Math.round(equityValue * 100) / 100
      },
      sensitivity,
      confidence: 'medium'
    };
  }

  calculateDDMValuation(data: Partial<ValuationInput>): ValuationResult {
    const dps = data.dividend_per_share || 0;
    const growth = (data.growth_rate || 2) / 100;
    const required = (data.required_return || 10) / 100;

    let value = 0;
    if (required > growth) {
      value = dps * (1 + growth) / (required - growth);
    }

    const sensitivity: Record<string, number> = {};
    [required - 0.02, required, required + 0.02].forEach(req => {
      [Math.max(0, growth - 0.01), growth, growth + 0.01].forEach(gr => {
        if (req > gr) {
          const val = dps * (1 + gr) / (req - gr);
          sensitivity[`R${(req * 100).toFixed(0)}%_G${(gr * 100).toFixed(0)}%`] = Math.round(val * 100) / 100;
        }
      });
    });

    return {
      method: 'DDM股利贴现模型',
      value_per_share: Math.round(value * 100) / 100,
      assumptions: {
        dividend_per_share: dps,
        growth_rate: `${(growth * 100).toFixed(0)}%`,
        required_return: `${(required * 100).toFixed(0)}%`,
        dividend_yield: value > 0 ? Math.round(dps / value * 10000) / 100 : 0
      },
      sensitivity,
      confidence: dps > 0 ? 'high' : 'low'
    };
  }

  comprehensiveValuation(data: Partial<ValuationInput>): {
    individual_results: ValuationResult[];
    valuation_range: { low: number; high: number; median: number; average: number } | null;
    timestamp: string;
  } {
    const results: ValuationResult[] = [];

    if (data.eps && data.eps > 0) {
      results.push(this.calculatePEValuation(data));
    }

    if (data.book_value_per_share && data.book_value_per_share > 0) {
      results.push(this.calculatePBValuation(data));
    }

    if (data.fcf && data.fcf !== 0) {
      results.push(this.calculateDCFValuation(data));
    }

    if (data.ebitda && data.ebitda > 0) {
      results.push(this.calculateEVEBITDAValuation(data));
    }

    if (data.dividend_per_share && data.dividend_per_share > 0) {
      results.push(this.calculateDDMValuation(data));
    }

    let valuationRange: { low: number; high: number; median: number; average: number } | null = null;

    if (results.length > 0) {
      const values = results
        .map(r => r.value_per_share || r.target_price)
        .filter((v): v is number => v !== undefined && v > 0);

      if (values.length > 0) {
        const sorted = [...values].sort((a, b) => a - b);
        valuationRange = {
          low: Math.round(sorted[0] * 100) / 100,
          high: Math.round(sorted[sorted.length - 1] * 100) / 100,
          median: Math.round(sorted[Math.floor(sorted.length / 2)] * 100) / 100,
          average: Math.round((sorted.reduce((a, b) => a + b, 0) / sorted.length) * 100) / 100
        };
      }
    }

    return {
      individual_results: results,
      valuation_range: valuationRange,
      timestamp: new Date().toISOString()
    };
  }
}

export const valuationEngine = new ValuationEngine();
